tech_soln/real_time_dashboards.rst

```diff
@@ -0,0 +1,423 @@
+keep raw data for one day and 1-minute aggregations for one month.
+
+.. code-block:: plpgsql
+
+  -- another function for the master
+  CREATE FUNCTION expire_old_request_data() RETURNS void
+  AS $$
+    SET LOCAL citus.all_modification_commutative TO TRUE;
+    SELECT master_modify_multiple_shards(
+      'DELETE FROM http_request WHERE ingest_time < now() - interval ''1 day'';');
+    SELECT master_modify_multiple_shards(
+      'DELETE FROM http_request_1min WHERE ingest_time < now() - interval ''1 month'';');
+  END;
+  $$ LANGUAGE 'sql';
+
+.. NOTE::
+
+  The above function should be called every minute. You could do this by adding a crontab
+  entry on the master node:
+
+  .. code-block:: bash
+  
+    * * * * * psql -c "SELECT expire_old_request_data();"
+
+That's the basic architecture! We provided an architecture that ingests HTTP events and
+then rolls up these events into their pre-aggregated form. This way, you can both store
+raw events and also power your analytical dashboards with subsecond queries.
+
+The next sections extend upon the basic architecture and show you how to resolve questions
+which often appear.
+
+Approximate Distinct Counts
+---------------------------
+
+A common question in http analytics deals with :ref:`approximate distinct counts
+<count_distinct>`: How many unique visitors visited your site over the last month?
+Answering this question exactly requires storing the list of all previously-seen visitors
+in the rollup tables, a prohibitively large amount of data. A datatype called hyperloglog,
+or HLL, can answer the query approximately; it takes a surprisingly small amount of space
+to tell you approximately how many unique elements are in a set you pass it. Its accuracy
+can be adjusted. We'll use ones which, using only 1280 bytes, will be able to count up to
+tens of billions of unique visitors with at most 2.2% error.
+
+An equivalent problem appears if you want to run a global query, such as the number of
+unique ip addresses which visited any of your client's sites over the last month. Without
+HLLs this query involves shipping lists of ip addresses from the workers to the master for
+it to deduplicate. That's both a lot of network traffic and a lot of computation. By using
+HLLs you can greatly improve query speed.
+
+First you must install the hll extension; `the github repo
+<https://github.com/aggregateknowledge/postgresql-hll>`_ has instructions. Next, you have
+to enable it:
+
+.. code-block:: sql
+
+  -- this part must be run on all nodes
+  CREATE EXTENSION hll;
+
+  -- this part runs on the master
+  ALTER TABLE http_request_1min ADD COLUMN distinct_ip_addresses (hll);
+
+When doing our rollups, we can now aggregate sessions into an hll column with queries
+like this:
+
+.. code-block:: sql
+
+  SELECT
+    site_id, date_trunc('minute', ingest_time) as minute,
+    hll_add_agg(hll_hash_text(ip_address)) AS distinct_ip_addresses
```

Comment:
Maybe we should write whole rollup function here instead of just new addition and then we can highlight new line to make it clear that it is mostly same except that hll line. In that way;

-This would be full example and it would be more clear that this is our new rollup function.
-A user who follows this document step by step, would have full version of the new function and can continue to play without needing to do any changes.